src/Sylius/Bundle/CoreBundle/Checkout/CheckoutResolver.php

```diff
@@ -0,0 +1,130 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Bundle\CoreBundle\Checkout;
+
+use SM\Factory\FactoryInterface;
+use Sylius\Component\Cart\Context\CartContextInterface;
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Core\OrderCheckoutTransitions;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestMatcherInterface;
+use Symfony\Component\HttpKernel\Event\GetResponseEvent;
+use Symfony\Component\HttpKernel\KernelEvents;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+
+/**
+ * @author Arkadiusz Krakowiak <arkadiusz.krakowiak@lakion.com>
+ */
+final class CheckoutResolver implements EventSubscriberInterface
+{
+    /**
+     * @var CartContextInterface
+     */
+    private $cartContext;
+
+    /**
+     * @var UrlGeneratorInterface
+     */
+    private $urlGenerator;
+
+    /**
+     * @var RequestMatcherInterface
+     */
+    private $requestMatcher;
+
+    /**
+     * @var FactoryInterface
+     */
+    private $stateMachineFactory;
+
+    /**
+     * @param CartContextInterface $cartContext
+     * @param UrlGeneratorInterface $urlGenerator
+     * @param RequestMatcherInterface $requestMatcher
+     * @param FactoryInterface $stateMachineFactory
+     */
+    public function __construct(
+        CartContextInterface $cartContext,
+        UrlGeneratorInterface $urlGenerator,
+        RequestMatcherInterface $requestMatcher,
+        FactoryInterface $stateMachineFactory
+    ) {
+        $this->cartContext = $cartContext;
+        $this->urlGenerator = $urlGenerator;
+        $this->requestMatcher = $requestMatcher;
+        $this->stateMachineFactory = $stateMachineFactory;
+    }
+
+    /**
+     * @param GetResponseEvent $event
+     */
+    public function onKernelRequest(GetResponseEvent $event)
+    {
+        if (!$event->isMasterRequest()) {
+            return;
+        }
+
+        $request = $event->getRequest();
+
+        if (!$this->requestMatcher->matches($request)) {
+            return;
+        }
+
+        /** @var OrderInterface $order */
+        $order = $this->cartContext->getCart();
+        $stateMachine = $this->stateMachineFactory->get($order, OrderCheckoutTransitions::GRAPH);
+
+        if ($stateMachine->can($this->getRequestedTransition($request))) {
+            return;
+        }
+
+        if (null !== $referer = $this->getReferer($request)) {
```

Comment:
I think this method call is unnecessary. We can skip it and simply write `$request->headers->get('referer')` here.